sklearn/compose/_column_transformer.py

```diff
@@ -246,13 +247,39 @@ def _iter(self, fitted=False, replace_strings=False):
                 elif _is_empty_column_selection(column):
                     continue
 
-            yield (name, trans, column, get_weight(name))
+            yield (name, column, trans, get_weight(name))
+
+    def _validate_tuple_order(self):
+        # check if any given tuple follows the
+        # (name, transformer, column) order.
+        # flip all tuples if that's the case.
+        # remove in v0.22
+        warn_message = ('ColumnTransformer transformer tuples should be '
+                        '(name, column, transformer), whereas '
+                        '(name, transformer, column) was passed; '
+                        'Assuming the deprecated (name, transformer, column) '
+                        'order for all given tuples. '
+                        'The support for the old order is deprecated in 0.20.1'
+                        ' and will be removed in version 0.22.')
+        try:
+            self._validate_transformers()
+        except TypeError:
+            try:
+                self.transformers = [(x, z, y)
```

Comment:
We avoid modifying the user's parameters when validating. Rather, we should make this a `_check_*` method and return the validated form to use locally, I think. WDYT, @amueller? Is it worth being pedantic about the don't-change-params requirement in this case?

Reply:
But complexity might come in for example in the following case

```
ct = ColumnTransformer([('trans', StandardScaler(), ['a', 'b'])])
ct.set_params(trans=MinMaxScaler())
```

Should we then also update the `ct.transformers` attribute in its *deprecated* order? So that means we need to keep track of which order is used, and then update the attribute accordingly.

Reply:
I think changing the internal representation will give us trouble to no end and I'd argue that maybe we should change ``make_column_transformer`` after all, given the added complexity we're getting here (and I need to change my slides *again* lol).

Reply:
If @jnothman wasn't listening, I'd say this was release as _experimental_ anyway, so we'd be allowed to be backward incompatible, with an error message, instead of a warning, which would tell users how they should change the code. But I guess Joel really would't like the approach.

Reply:
I would personally not come back (again) on our earlier decision, just for what we are discussing here. 
I would say to either decide we can exceptionally change the user's parameter in this specific case, or either just live with the complexity. What I said above will make the code more complex, but it is not undo-able, and will only be temporarily.

Reply:
Changing the user's parameter doesn't solve the ``set_params`` issue, does it? We would still need to detect which way it is stored and then overwrite accordingly, right?

Reply:
Yes, but that is "just" one more private attribute